src/krb5_plugin/idp/idp_kdcpreauth.c

```diff
@@ -0,0 +1,807 @@
+                          krb5_kdcpreauth_callbacks cb,
+                          krb5_kdcpreauth_rock rock,
+                          krb5_kdcpreauth_edata_respond_fn respond,
+                          void *arg,
+                          struct sss_idpkdc_config *config)
+{
+    struct sss_idpkdc_challenge *state;
+    krb5_error_code ret;
+
+    state = malloc(sizeof(struct sss_idpkdc_challenge));
+    if (state == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+    memset(state, 0, sizeof(struct sss_idpkdc_challenge));
+
+    state->kctx = kctx;
+    state->cb = cb;
+    state->rock = rock;
+    state->respond = respond;
+    state->arg = arg;
+
+    state->radius = sss_idpkdc_radius_init(kctx, vctx, config);
+    if (state->radius == NULL) {
+        ret = ENOMEM;
+        goto done;
+    }
+
+    ret = krad_client_send(state->radius->client,
+                           krad_code_name2num("Access-Request"),
+                           state->radius->attrs, config->server,
+                           config->secret, config->timeout, config->retries,
+                           sss_idpkdc_challenge_done, state);
+
+done:
+    if (ret != 0) {
+        sss_idpkdc_challenge_free(state);
+    }
+
+    return ret;
+}
+
+struct sss_idpkdc_verify {
+    struct sss_idpkdc_radius *radius;
+    struct sss_idpkdc_config *config;
+
+    krb5_context kctx;
+    krb5_kdcpreauth_rock rock;
+    krb5_kdcpreauth_callbacks cb;
+    krb5_enc_tkt_part *enc_tkt_reply;
+    krb5_kdcpreauth_verify_respond_fn respond;
+    void *arg;
+};
+
+static void
+sss_idpkdc_verify_free(struct sss_idpkdc_verify *verify)
+{
+    if (verify == NULL) {
+        return;
+    }
+
+    sss_idpkdc_radius_free(verify->radius);
+    sss_idpkdc_config_free(verify->config);
+    free(verify);
+}
+
+static void
+sss_idpkdc_verify_done(krb5_error_code rret,
+                       const krad_packet *rreq,
+                       const krad_packet *rres,
+                       void *data)
+{
+    static bool verify_success = true;
+    static bool verify_failure = false;
+    struct sss_idpkdc_verify *state;
+    krb5_kdcpreauth_modreq modreq;
+    krb5_error_code ret;
+    int i;
+
+    state = (struct sss_idpkdc_verify *)data;
+    modreq = (krb5_kdcpreauth_modreq)&verify_failure;
+
+    if (rret != 0) {
+        ret = rret;
+        goto done;
+    }
+
+    if (krad_packet_get_code(rres) != krad_code_name2num("Access-Accept")) {
+        ret = KRB5_PREAUTH_FAILED;
+        goto done;
+    }
+
+    state->enc_tkt_reply->flags |= TKT_FLG_PRE_AUTH;
+
+    for (i = 0; state->config->idpcfg->indicators[i] != NULL; i++) {
+        ret = state->cb->add_auth_indicator(state->kctx, state->rock,
+                                            state->config->idpcfg->indicators[i]);
+        if (ret != 0) {
+            goto done;
+        }
+    }
+
+    modreq = (krb5_kdcpreauth_modreq)&verify_success;
+
+done:
+    state->respond(state->arg, ret, modreq, NULL, NULL);
```

Comment:
Coverity:
```
[Error: CLANG_WARNING: ]()[[#def2]](https://cov01.lab.eng.brq.redhat.com/covscanhub/task/242632/log/added.html#def2)
sssd-2.6.2/src/krb5_plugin/idp/idp_kdcpreauth.c:509:5: warning[core.CallAndMessage]: 2nd function call argument is an uninitialized value
#  507|   
#  508|   done:
#  509|->     state->respond(state->arg, ret, modreq, NULL, NULL);
#  510|       sss_idpkdc_verify_free(state);
#  511|       return;
```

If everything goes well and there are no indicators `ret` will not be set.